# 배경
- Sepsis 환자의 다양한 정보를 활용하여 다음 방문에 처방할 약을 맞출수 있을까? 
- [TBD] 처방 받는 적절한 이유가 필요함.


# 문제 종류
- binary classification: 항생제를 처방받았는가
- [TBD] multi-label classification: 어떤 항생제를 처방받았는가


# 실험
- 코호트

<img src="images/pnp_fig1.png">

- 피처 전처리/엔지니어링
    - [TBD]
- 모델
    - LR
    - RF
    - GB
    - (R)NN

# 교차검증
- 5-folds CV

# 평가
- AUROC
- variable importance: 어떤 변수가 다음 처방에 영향을 미쳤을까?


# 예측값의 활용
- 처방 실수의 예방

In [4]:
%matplotlib inline

import matplotlib.pyplot as plt
import seaborn as sns; sns.set(rc={'figure.figsize':(15,15)})
import numpy as np
import pandas as pd
from sqlalchemy import create_engine
engine = create_engine('postgresql://postgres:mimic@127.0.0.1:5555/mimic')

In [42]:
sql = """
select a.subject_id, s.* 
from sepsis3_cohort s 
    left join icustays a 
    on s.hadm_id=a.hadm_id and s.icustay_id=a.icustay_id
"""
label = pd.read_sql(sql, engine)
label.head().T

0                    1  \
subject_id                                        55973                27513   
hadm_id                                          152234               163557   
icustay_id                                       200001               200003   
intime                              2181-11-25 19:06:12  2199-08-02 19:50:04   
outtime                             2181-11-28 20:59:25  2199-08-08 17:09:18   
age                                             61.0499               48.294   
gender                                                F                    M   
ethnicity                          ASIAN - ASIAN INDIAN                WHITE   
first_service                                       MED                 SURG   
dbsource                                     metavision              carevue   
suspected_of_infection_poe                            1                    1   
suspected_infection_time_poe        2181-11-18 11:10:00  2199-08-02 21:02:00   
suspected_infection_time_poe_days               7.33069           -0.0499537   
specimen_poe                              BLOOD CULTURE        BLOOD CULTURE   
positiveculture_poe                                   0                    1   
antibiotic_time_poe                 2181-11-18 00:00:00  2199-08-02 00:00:00   
exclusion_secondarystay                               1                    0   
exclusion_nonadult                                    0                    0   
exclusion_csurg                                       0                    0   
exclusion_carevue                                     0                    1   
exclusion_early_suspicion                             1                    0   
exclusion_late_suspicion                              0                    0   
exclusion_bad_data                                    0                    0   
excluded                                              1                    1   

                                                     2                    3  \
subject_id                                       10950                20707   
hadm_id                                         189514               129310   
icustay_id                                      200006               200007   
intime                             2159-09-03 11:28:14  2109-02-17 10:03:37   
outtime                            2159-09-04 19:08:10  2109-02-18 17:03:12   
age                                             54.071              43.3439   
gender                                               M                    M   
ethnicity                                        OTHER                WHITE   
first_service                                      MED                 CMED   
dbsource                                       carevue              carevue   
suspected_of_infection_poe                           0                    0   
suspected_infection_time_poe                       NaT                  NaT   
suspected_infection_time_poe_days                  NaN                  NaN   
specimen_poe                                      None                 None   
positiveculture_poe                                NaN                  NaN   
antibiotic_time_poe                                NaT                  NaT   
exclusion_secondarystay                              1                    0   
exclusion_nonadult                                   0                    0   
exclusion_csurg                                      0                    0   
exclusion_carevue                                    1                    1   
exclusion_early_suspicion                            0                    0   
exclusion_late_suspicion                             0                    0   
exclusion_bad_data                                   0                    0   
excluded                                             1                    1   

                                                     4  
subject_id          

In [29]:
label.groupby('excluded').size()

excluded
0    11791
1    49741
dtype: int64

In [30]:
label.subject_id.nunique()

46476

In [31]:
label[label.excluded==0].subject_id.nunique()

11791

In [32]:
label[label.excluded==1].subject_id.nunique()

36839

In [33]:
sql = """
select count(distinct subject_id)
from icustays
"""
pd.read_sql(sql, engine)

count
0  46476

- icu의 총 환자 46476명 중에 sepsis3 조건에 맞는 icustay는 11791 건
    - 해당 조건의 고유 환자수도 11791건으로 icustays:subejct_id = 1:1
    - 즉, 각 환자가 한번만 icu에 입원했고, sepsis 진단을 받았다. 여러번 입원하고 진단받은 사례는 없다.
    - 따라서, 환자 수준의 모델링만 가능, 방문 수준의 모델링은 불가능

<hr>

In [40]:
label[(label.excluded==0) & (label.antibiotic_time_poe.notnull())]['icustay_id'].count()

7588

In [41]:
label[(label.excluded==0) & (label.antibiotic_time_poe.isnull())]['icustay_id'].count()

4203

In [43]:
label[(label.excluded==0) & (label.antibiotic_time_poe.notnull())].subject_id.nunique()

7588

In [44]:
label[(label.excluded==0) & (label.antibiotic_time_poe.isnull())].subject_id.nunique()

4203

- icu의 총 환자 46476명 중에 sepsis3 조건에 맞는 icustay는 11791 건
    - 항생제 처방을 받은 환자수 7588, 건수 7588, 
    - 항생제 처방을 받지 않은 환자수 4203, 건수는 4203

<hr>

In [47]:
sql = """
select i.subject_id, s.*, a.antibiotic_time_poe
from sepsis3 s 
    join sepsis3_cohort a 
        on s.icustay_id=a.icustay_id
    join icustays i
        on s.icustay_id=i.icustay_id
where a.excluded = 0
"""
features = pd.read_sql(sql, engine)

In [48]:
features.head().T

0                    1  \
subject_id                                       61691                41710   
icustay_id                                      200021               200028   
hadm_id                                         109307               181955   
excluded                                             0                    0   
intime                             2114-12-26 19:45:12  2133-10-29 17:13:50   
outtime                            2114-12-27 22:46:28  2133-11-01 14:55:14   
dbsource                                    metavision           metavision   
suspected_infection_time_poe                       NaT                  NaT   
suspected_infection_time_poe_days                  NaN                  NaN   
specimen_poe                                      None                 None   
positiveculture_poe                                NaN                  NaN   
antibiotic_time_poe                                NaT                  NaT   
blood_culture_time                 2114-12-27 15:00:00  2133-10-29 19:27:00   
blood_culture_positive                               0                    0   
age                                            60.8501              64.8666   
gender                                               M                    M   
is_male                                              1                    1   
ethnicity                                        WHITE                WHITE   
race_white                                           1                    1   
race_black                                           0                    0   
race_hispanic                                        0                    0   
race_other                                           0                    0   
metastatic_cancer                                    0                    0   
diabetes                                             0                    1   
height                                             NaN               167.64   
weight                                            82.6                   84   
bmi                                                NaN              29.8899   
first_service                                     NMED                 CMED   
hospital_expire_flag                                 0                    0   
thirtyday_expire_flag                                0                    0   
icu_los                                         1.1259               2.9038   
hosp_los                                       1.94861              3.20417   
sepsis_angus                                         0                    0   
sepsis_martin                                        0                    0   
sepsis_explicit                                      0                    0   
septic_shock_explicit                                0                    0   
severe_sepsis_explicit                               0                    0   
sepsis_nqf                                           0                    0   
sepsis_cdc                                           0                    0   
sepsis_cdc_simple                                    0                    0   
elixhauser_hospital                                  4                    8   
vent                                                 1                    1   
sofa                                                 4                   10   
lods                                                 7                    6   
sirs                                                 3                    3   
qsofa                                                3                    2   
qsofa_sysbp_score                                    1                    1   
qsofa_gcs_score                                      1                    0   
qsofa_resprate_score                                 1                    1   
antibiotic_time_poe                                NaT                  NaT   

                       

# TBD

- 위의 피처 전처리를 어떻게 할것인가
- 추가로 붙일 피처가 있을까


# Reference 

[1] Sepsis-3 in MIMIC-III, https://github.com/alistairewj/sepsis3-mimic